controllers/reconcilers/grafana/grafana_service_reconciler.go

```diff
@@ -54,6 +54,24 @@ func (r *ServiceReconciler) Reconcile(ctx context.Context, cr *v1beta1.Grafana,
 			int32(GetGrafanaPort(cr))) // #nosec G115
 	}
 
+	// Headless service for grafana unified alerting
+	headlessService := model.GetGrafanaHeadlessService(cr, scheme)
+	_, err = controllerutil.CreateOrUpdate(ctx, r.client, headlessService, func() error {
+		model.SetCommonLabels(service)
+		service.Spec = v1.ServiceSpec{
+			ClusterIP: "None",
+			Ports:     getHeadlessServicePorts(cr),
```

Comment:
This function is only called on update, which is why the e2e tests are currently failing. You'll need to specify the entire object in `GetGrafanaHeadlessService` as well.